### ISSUE TITLE
Field Trait: Consider adding static linear combination operation

### ISSUE THREAD

**Issue by ValarDragon** [arkworks-rs/algebra#121]
SNARK Protocols (in particular verifiers) often have expensive linear combinations they have to evaluate. The most expensive term in a field multiplication is the modular reduction. However, if we're doing a linear combination, we should be able to batch the modular reductions by adding all of the products in their un-reduced forms.

e.g. if we have an M limb field, and we want to compute `c = sum_{i in 0..N} a_i * b_i`, we can compute this as (pseudocode ignoring montgomery reductions):
```
let mut unreduced_c: 2M + 1 limb bigint;
for i in 0..N {
    let term: 2M limb bigint = unreduced_product(a_i, b_i);
    unreduced_c += term;
}
let c = modular_reduce(c);
c
```
We're guaranteed unreduced_c won't overflow if N < 2^64, since it has one extra limb.

To handle that we're in montgomery form, as a first method, we could make `modular_reduce` just reduce by `p` and then multiply by `R'` at the end. I think its possible to adapt REDC to combine the final multiplication by `R'` into the reduction, but the speedup of that should be marginal, and likely has much higher code complexity.

This could potentially be a significant speedup.

**Comment by Pratyush** [comment]
I imagine this is more useful for something like Aurora or Fractal? Because I don't think it's a bottleneck in EC-based protocols

**Comment by ValarDragon** [comment]
oh that might be the case. This was what the bottleneck was in Aurora's verifier. (And anything thats bottlenecked by lagrange interpolation) I wonder how important it is in witness generation

**Comment by ValarDragon** [comment]
I wonder if this would also be helpful as a sub-routine within making a batch addition formulae for elliptic curve points. (Not sure where that would be costly within a SNARK outside of a protocol engineered to use binary MSMs)

**Comment by ValarDragon** [comment]
BLST's implementation reports a 15% speedup in their sqrt code by using unreduced squarings: https://github.com/supranational/blst/blob/194c2623799416c76edf6e62fcabdc59aa1e9f80/src/sqrt.c#L23

**Comment by Pratyush** [comment]
Oh that's quite a good improvement, seems like this API is definitely justified. I presume we can add a `linear_combination` method on `PrimeField` that has a default implementation which just does the `sum_i(a_i * b_i)`, and then we can add separate optimized impls later.

**Comment by ValarDragon** [comment]
We have further evidence from #80 that actually a more extreme version of this is an efficiency gain. The more extreme version namely being returning the unreduced form directly. There it was shown that the fp12 arithmetic was half the time as what we currently have, despite having same if not slightly more performant base field arithmetic. Similarly, fp2 arithmetic was ~25% faster.

**Comment by jon-chuang** [comment]
I'm very much in favour of this. Reducing the number of reduce operations in montgomery mul can make huge savings, possible on the order of 1/3 - 1/2, which is about how much we are lagging right now.

Unfortunately, the number of additions/subtractions one can perform before reducing is field dependent. However, we can count leading zeroes and use a constant branch into the right impl. While messy, I think it could be very promising.

**Comment by Pratyush** [comment]
This can have an impact in our elliptic curve formulae as well, as there are a number of places where we compute a chain of additions. I'm thinking of two APIs:

```rust
fn linear_combination(&[(Self, Self)]) -> Self;

pub enum Op {
	Add,
	Sub,
}

impl Sum<(Op, Self)> for Fp {
// compute a running sum while avoiding reductions.
}
```

**Comment by ValarDragon** [comment]
Is `Sum` the right trait? Seems to require queuing an iterator which will take much more memory (unless I misunderstand the trait https://doc.rust-lang.org/std/iter/trait.Sum.html )

**Comment by Pratyush** [comment]
You can construct the iterator lazily via a closure, so it depends entirely on how the iterator is constructed.

**Comment by Pratyush** [comment]
Will be closed by #418
